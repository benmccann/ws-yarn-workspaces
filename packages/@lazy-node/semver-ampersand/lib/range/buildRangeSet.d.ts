import { IOptions } from '../types';
export declare function buildRangeSet(range: string, options?: IOptions): string[][];
