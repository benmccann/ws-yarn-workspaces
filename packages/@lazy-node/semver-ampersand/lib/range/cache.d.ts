import { LRUCache } from 'lru-cache2';
import { Comparator } from 'semver';
export declare const cache: LRUCache<string, readonly Comparator[]>;
