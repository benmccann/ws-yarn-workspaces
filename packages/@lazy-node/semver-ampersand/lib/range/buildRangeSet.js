"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildRangeSet = void 0;
const split_1 = require("../util/split");
const array_hyper_unique_1 = require("array-hyper-unique");
const handleAmpersandAndSpaces_1 = require("../handleAmpersandAndSpaces");
const normalizeRangeInput_1 = require("./normalizeRangeInput");
function buildRangeSet(range, options = {}) {
    range = (0, handleAmpersandAndSpaces_1.handleAmpersandAndSpaces)(range, options);
    let rangeSet = (0, split_1.splitDoubleVerticalBar)(range)
        // map the range to a 2d array of comparators
        .map(range => {
        range = (0, normalizeRangeInput_1.normalizeRangeInput)(range, options);
        return (0, split_1.splitSpace)(range);
    });
    rangeSet = (0, array_hyper_unique_1.array_unique_overwrite)(rangeSet);
    return rangeSet;
}
exports.buildRangeSet = buildRangeSet;
//# sourceMappingURL=buildRangeSet.js.map