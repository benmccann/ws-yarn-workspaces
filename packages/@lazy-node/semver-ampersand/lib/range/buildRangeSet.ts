import { IOptions } from '../types';
import { splitDoubleVerticalBar, splitSpace } from '../util/split';
import { array_unique_overwrite } from 'array-hyper-unique';
import { handleAmpersandAndSpaces } from '../handleAmpersandAndSpaces';
import { normalizeRangeInput } from './normalizeRangeInput';

export function buildRangeSet(range: string, options: IOptions = {})
{
	range = handleAmpersandAndSpaces(range, options);

	let rangeSet = splitDoubleVerticalBar(range)
		// map the range to a 2d array of comparators
		.map(range =>
		{
			range = normalizeRangeInput(range, options);

			return splitSpace(range)
		})
	;

	rangeSet = array_unique_overwrite(rangeSet)

	return rangeSet
}
