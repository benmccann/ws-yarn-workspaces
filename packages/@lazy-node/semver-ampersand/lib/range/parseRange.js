"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseRange = exports.reduceComparatorList = exports.filterRangeListForComparator = exports.parseRangeCore = exports.normalizeRangeInputForComparator = exports.getMemoOpts = void 0;
const cache_1 = require("./cache");
const semver_1 = require("semver");
const detect_1 = require("../comparator/detect");
const re_1 = require("semver/internal/re");
const util_1 = require("./util");
const split_1 = require("../util/split");
const array_hyper_unique_1 = require("array-hyper-unique");
const normalizeRangeInput_1 = require("./normalizeRangeInput");
/**
 * memoize range parsing for performance.
 * this is a very hot path, and fully deterministic.
 */
function getMemoOpts(options) {
    return Object.keys(options).filter(k => options[k]).join(',');
}
exports.getMemoOpts = getMemoOpts;
function normalizeRangeInputForComparator(range, options) {
    let rangeList = (0, split_1.splitSpace)(range)
        .map(comp => (0, util_1.parseComparator)(comp, options));
    rangeList = (0, split_1.splitSpace)(rangeList.join(' '))
        // >=0.0.0 is equivalent to *
        .map(comp => (0, util_1.replaceGTE0)(comp, options));
    return (0, array_hyper_unique_1.array_unique_overwrite)(rangeList);
}
exports.normalizeRangeInputForComparator = normalizeRangeInputForComparator;
function parseRangeCore(range, options) {
    range = (0, normalizeRangeInput_1.normalizeRangeInput)(range, options);
    // At this point, the range is completely trimmed and
    // ready to be split into comparators.
    let rangeList = normalizeRangeInputForComparator(range, options);
    if (options.loose) {
        rangeList = filterRangeListForComparator(rangeList, options);
    }
    let compList = rangeList
        .map(comp => new semver_1.Comparator(comp, options));
    const result = reduceComparatorList(compList);
    return result;
}
exports.parseRangeCore = parseRangeCore;
/**
 * in loose mode, throw out any that are not valid comparators
 */
function filterRangeListForComparator(rangeList, options) {
    const compRe = options.loose ? re_1.re[re_1.t.COMPARATORLOOSE] : re_1.re[re_1.t.COMPARATOR];
    return rangeList.filter(comp => !!comp.match(compRe));
}
exports.filterRangeListForComparator = filterRangeListForComparator;
/**
 * if any comparators are the null set, then replace with JUST null set
 * if more than one comparator, remove any * comparators
 * also, don't include the same comparator more than once
 */
function reduceComparatorList(compList) {
    const l = compList.length;
    const rangeMap = new Map();
    for (const comp of compList) {
        if ((0, detect_1.isNullSet)(comp)) {
            return [comp];
        }
        rangeMap.set(comp.value, comp);
    }
    if (rangeMap.size > 1 && rangeMap.has('')) {
        rangeMap.delete('');
    }
    const result = [...rangeMap.values()];
    return result;
}
exports.reduceComparatorList = reduceComparatorList;
function parseRange(range, options) {
    range = range.trim();
    // memoize range parsing for performance.
    // this is a very hot path, and fully deterministic.
    const memoOpts = getMemoOpts(options);
    const memoKey = `parseRange:${memoOpts}:${range}`;
    const cached = cache_1.cache.get(memoKey);
    if (cached) {
        return cached;
    }
    const result = parseRangeCore(range, options);
    cache_1.cache.set(memoKey, result);
    return result;
}
exports.parseRange = parseRange;
//# sourceMappingURL=parseRange.js.map