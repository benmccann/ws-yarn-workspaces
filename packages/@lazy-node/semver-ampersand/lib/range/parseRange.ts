import { cache } from './cache';
import { Comparator } from 'semver';
import { isNullSet } from '../comparator/detect';
import { IOptions } from '../types';
import { re, t } from 'semver/internal/re'
import { parseComparator, replaceGTE0 } from './util';
import { splitSpace } from '../util/split';
import { array_unique_overwrite } from 'array-hyper-unique';
import { normalizeRangeInput } from './normalizeRangeInput';

/**
 * memoize range parsing for performance.
 * this is a very hot path, and fully deterministic.
 */
export function getMemoOpts(options: IOptions)
{
	return Object.keys(options).filter(k => options[k]).join(',')
}

export function normalizeRangeInputForComparator(range: string, options: IOptions)
{
	let rangeList = splitSpace(range)
		.map(comp => parseComparator(comp, options))
	;

	rangeList = splitSpace(rangeList.join(' '))
		// >=0.0.0 is equivalent to *
		.map(comp => replaceGTE0(comp, options))

	return array_unique_overwrite(rangeList)
}

export function parseRangeCore(range: string, options: IOptions): ReadonlyArray<Comparator>
{
	range = normalizeRangeInput(range, options);

	// At this point, the range is completely trimmed and
	// ready to be split into comparators.

	let rangeList = normalizeRangeInputForComparator(range, options)

	if (options.loose)
	{
		rangeList = filterRangeListForComparator(rangeList, options)
	}

	let compList = rangeList
		.map(comp => new Comparator(comp, options))

	const result = reduceComparatorList(compList)

	return result
}

/**
 * in loose mode, throw out any that are not valid comparators
 */
export function filterRangeListForComparator(rangeList: string[], options: IOptions)
{
	const compRe = options.loose ? re[t.COMPARATORLOOSE] : re[t.COMPARATOR]

	return rangeList.filter(comp => !!comp.match(compRe))
}

/**
 * if any comparators are the null set, then replace with JUST null set
 * if more than one comparator, remove any * comparators
 * also, don't include the same comparator more than once
 */
export function reduceComparatorList(compList: Comparator[])
{
	const l = compList.length
	const rangeMap = new Map<string, Comparator>()
	for (const comp of compList)
	{
		if (isNullSet(comp))
		{
			return [comp]
		}
		rangeMap.set(comp.value, comp)
	}
	if (rangeMap.size > 1 && rangeMap.has(''))
	{
		rangeMap.delete('')
	}

	const result = [...rangeMap.values()]

	return result
}

export function parseRange(range: string, options: IOptions): ReadonlyArray<Comparator>
{
	range = range.trim()

	// memoize range parsing for performance.
	// this is a very hot path, and fully deterministic.
	const memoOpts = getMemoOpts(options)
	const memoKey = `parseRange:${memoOpts}:${range}`
	const cached = cache.get(memoKey)
	if (cached)
	{
		return cached
	}

	const result = parseRangeCore(range, options)

	cache.set(memoKey, result)
	return result
}
