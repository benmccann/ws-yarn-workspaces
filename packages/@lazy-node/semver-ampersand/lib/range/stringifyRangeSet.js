"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stringifyRangeSet = void 0;
function stringifyRangeSet(rangeSet) {
    return rangeSet.flat().join(' ');
}
exports.stringifyRangeSet = stringifyRangeSet;
//# sourceMappingURL=stringifyRangeSet.js.map