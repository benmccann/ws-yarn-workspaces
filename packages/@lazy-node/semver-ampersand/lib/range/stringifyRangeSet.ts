import { buildRangeSet } from './buildRangeSet';

export function stringifyRangeSet(rangeSet: ReturnType<typeof buildRangeSet>)
{
	return rangeSet.flat().join(' ')
}
