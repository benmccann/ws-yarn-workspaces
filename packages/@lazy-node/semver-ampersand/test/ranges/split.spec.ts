import { filterRangeListForComparator } from '../../lib/range/parseRange';
import { buildRangeSet } from '../../lib/range/buildRangeSet';
import { stringifyRangeSet } from '../../lib/range/stringifyRangeSet';

describe(`describe`, () =>
{

	let input = '>x 2.x || * || <x || ~1.2.3beta ^4.0.0-dev.20200615 && 4.0.0-dev.20200615 < 4.0.0-dev.20200800 || 4.0.x - 4.0.2 || (4.0.x - 4.0.2) || >= 1 || =1 || ~> 2.0.0 || 1.2.3 - 1.2.4 - 1.2.5 || >=1.2.3 <=1.2.4';

	test(input, () =>
	{

		let actual = buildRangeSet(input, {
			loose: true,
		});

		expect(actual).toMatchSnapshot();

		let actual2 = buildRangeSet(input, {
			loose: false,
		});

		expect(actual2).toMatchSnapshot();

		expect(actual).toStrictEqual(actual2)

		expect(stringifyRangeSet(actual2)).toMatchSnapshot();
	});

})
