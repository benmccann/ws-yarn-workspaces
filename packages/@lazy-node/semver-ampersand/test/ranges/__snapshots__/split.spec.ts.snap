// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe >x 2.x || * || <x || ~1.2.3beta ^4.0.0-dev.20200615 && 4.0.0-dev.20200615 < 4.0.0-dev.20200800 || 4.0.x - 4.0.2 || (4.0.x - 4.0.2) || >= 1 || =1 || ~> 2.0.0 || 1.2.3 - 1.2.4 - 1.2.5 || >=1.2.3 <=1.2.4 1`] = `
Array [
  Array [
    ">x",
    "2.x",
  ],
  Array [
    "*",
  ],
  Array [
    "<x",
  ],
  Array [
    "~1.2.3beta",
    "^4.0.0-dev.20200615",
    "4.0.0-dev.20200615",
    "<4.0.0-dev.20200800",
  ],
  Array [
    ">=4.0.0",
    "<=4.0.2",
  ],
  Array [
    "(4.0.x - 4.0.2)",
  ],
  Array [
    ">=1",
  ],
  Array [
    "=1",
  ],
  Array [
    "~>2.0.0",
  ],
  Array [
    "1.2.3",
    "-",
    "1.2.4",
    "-",
    "1.2.5",
  ],
  Array [
    ">=1.2.3",
    "<=1.2.4",
  ],
]
`;

exports[`describe >x 2.x || * || <x || ~1.2.3beta ^4.0.0-dev.20200615 && 4.0.0-dev.20200615 < 4.0.0-dev.20200800 || 4.0.x - 4.0.2 || (4.0.x - 4.0.2) || >= 1 || =1 || ~> 2.0.0 || 1.2.3 - 1.2.4 - 1.2.5 || >=1.2.3 <=1.2.4 2`] = `
Array [
  Array [
    ">x",
    "2.x",
  ],
  Array [
    "*",
  ],
  Array [
    "<x",
  ],
  Array [
    "~1.2.3beta",
    "^4.0.0-dev.20200615",
    "4.0.0-dev.20200615",
    "<4.0.0-dev.20200800",
  ],
  Array [
    ">=4.0.0",
    "<=4.0.2",
  ],
  Array [
    "(4.0.x - 4.0.2)",
  ],
  Array [
    ">=1",
  ],
  Array [
    "=1",
  ],
  Array [
    "~>2.0.0",
  ],
  Array [
    "1.2.3",
    "-",
    "1.2.4",
    "-",
    "1.2.5",
  ],
  Array [
    ">=1.2.3",
    "<=1.2.4",
  ],
]
`;

exports[`describe >x 2.x || * || <x || ~1.2.3beta ^4.0.0-dev.20200615 && 4.0.0-dev.20200615 < 4.0.0-dev.20200800 || 4.0.x - 4.0.2 || (4.0.x - 4.0.2) || >= 1 || =1 || ~> 2.0.0 || 1.2.3 - 1.2.4 - 1.2.5 || >=1.2.3 <=1.2.4 3`] = `">x 2.x * <x ~1.2.3beta ^4.0.0-dev.20200615 4.0.0-dev.20200615 <4.0.0-dev.20200800 >=4.0.0 <=4.0.2 (4.0.x - 4.0.2) >=1 =1 ~>2.0.0 1.2.3 - 1.2.4 - 1.2.5 >=1.2.3 <=1.2.4"`;
